### ISSUE TITLE
Start using the Regex source generator in this repo

### ISSUE THREAD

**Issue by danmoseley** [dotnet/runtime#62105]
We now have a source generator in .NET 7 SDK builds to generate regex code at compile time. We should start using it in this repo. (Not in the System.Text.RegularExpressions tests, of course - but in tests elsewhere in the repo there are some regexes.)

Up for grabs. If this works out we can open an issue in aspnetcore to do the same.

This needs main to start using the .NET 7 SDK first. @dotnet/runtime-infrastructure is that planned soon?

cc @stephentoub

**Comment by Clockwork-Muse** [comment]
It's been a while since I've done anything.  I suppose I can take this on (will keep me out of trouble).

**Comment by am11** [comment]
The naming difference in `[RegexGenerator]` and `[GeneratedDllImport]` (both corresponding to upcoming .NET 7 generators) might look a bit strange when working in common configurations like the ones in `eng/generators.targets`. It was mentioned under `Alternative names` in https://github.com/dotnet/runtime/issues/58880. It would be nice to align their names, but not sure if it is too late to propose name change; `GeneratedDllImport` to match regex one (or vice versa) before .NET 7 ships.

**Comment by stephentoub** [comment]
> not sure if it is too late to propose name change; GeneratedDllImport to match regex one (or vice versa) before .NET 7 ships.

It's not too late.  I agree we should be consistent between them.  Can you please open a separate issue?

(Personally, I prefer RegexGeneratorAttribute / DllImportGeneratorAttribute, as it's the thing causing the generation to happen rather than being the result of the generation process.)

**Comment by danmoseley** [comment]
> This needs main to start using the .NET 7 SDK first. @dotnet/runtime-infrastructure is that planned soon?

Seems that has now happened. https://github.com/dotnet/runtime/pull/61472

**Comment by jkoritzinsky** [comment]
Why don't we use the live generator instead of the LKG one from the SDK? That way we have a better loop for fixing bugs in the generator and we don't have to update to use a nightly unsigned SDK.

**Comment by danmoseley** [comment]
If that's technically possible (sounds like build phases all over again?) then I agree that's better.

**Comment by jkoritzinsky** [comment]
That's what we're doing with the DllImport source generator. We're using the same mechanisms used to test the generator to use them in the rest of the repo itself.

**Comment by danmoseley** [comment]
Perfect, then @Clockwork-Muse you should be unblocked. thanks for offering to do this!

We will also need to solve this problem for aspnetcore, or they will, if they want to opt into this in the near term. Let's do our repo first.

**Comment by jkoritzinsky** [comment]
If the Regex generator is shipping in the .NET 7 ref pack, then ASP.NET Core should be able to consume it by the same mechanism that they consume the nightly ref packs today.

**Comment by Clockwork-Muse** [comment]
Confirm that I can use the in-repo generator.

If there are non-test projects that are using regex, should I switch them as well?

**Comment by danmoseley** [comment]
> If there are non-test projects that are using regex, should I switch them as well?

I see no reason why not (there is hardly any usage there though) but we might want to do some perf measurements for some of those. I don't believe we expect them to be any slower but this is new.

**Comment by Clockwork-Muse** [comment]
Style question: Do we have a preferred naming convention for these?  Currently I've been doing something like:

```csharp
[RegexGenerator(@"^\d{4}\D\d{2}\D\d{2}\D\d{2}\D\d{2}\D\d{2}\D\d{2}\s$")]
private static partial Regex GetFormatterNameSimpleRegex();
```

**Comment by ericstj** [comment]
cc @joperezr 

> Why don't we use the live generator instead of the LKG one from the SDK?

Yes.  Here's a pointer: https://github.com/dotnet/runtime/blob/84b55d9700da70e3e68456929a4b49295772d87b/eng/generators.targets#L35

> If the Regex generator is shipping in the .NET 7 ref pack, then ASP.NET Core should be able to consume it by the same mechanism that they consume the nightly ref packs today.

Correct.  That was enabled with https://github.com/dotnet/runtime/pull/61329

**Comment by danmoseley** [comment]
> Do we have a preferred naming convention for these?

I think you might be helping figure that out! What you have seems reasonable - to me at least. I could imagine if it's used in several places the calling code might wrap it in a property so it's FormatterNameSimpleRegex.Match or whatnot rather than GetformatterNameSimpleRegex().Match.

**Comment by joperezr** [comment]
>I see no reason why not (there is hardly any usage there though) but we might want to do some perf measurements for some of those. I don't believe we expect them to be any slower but this is new.

For all of the benchmarks we have so far, source generator is much faster than the interpreted engine (which is the default). Only places I'd imagine this might be an issue is when hitting places that the source generator doesn't support yet, like using NonBacktracking engine for instance, but I really doubt any of our tests is doing that.

**Comment by stephentoub** [comment]
The source generator should be as fast or faster than RegexOptions.Compiled for everything; if it's not, it's a bug.  We should be comfortable switching anything to use the source generator from a throughput perspective.

**Comment by Clockwork-Muse** [comment]
.... also, because I can be terrible about this, how many PRs do you want?  One for everything?  There's ~600 files, but most of the changes are relatively minor.

**Comment by danmoseley** [comment]
Do you want to just put up the smallest possible one first so we see what it looks like?

**Comment by Clockwork-Muse** [comment]
@danmoseley #62325 

Only one regex to replace on a project (although it's a source one, not a test one).

**Comment by Clockwork-Muse** [comment]
... [replacing this feels wrong](https://github.com/dotnet/runtime/blob/fedf64f5688246b73505537c4dd81fc0afb57ddb/src/libraries/Microsoft.Extensions.Logging.Console/tests/Microsoft.Extensions.Logging.Console.Tests/JsonConsoleFormatterTests.cs#L277-L314):

```csharp
[ConditionalTheory(typeof(PlatformDetection), nameof(PlatformDetection.IsThreadingSupported))]
[MemberData(nameof(FloatingPointValues))]
public void Log_StateAndScopeContainsFloatingPointType_SerializesValue(object value)
{
    // Arrange
    var t = SetUp(
        new ConsoleLoggerOptions { FormatterName = ConsoleFormatterNames.Json },
        simpleOptions: null,
        systemdOptions: null,
        jsonOptions: new JsonConsoleFormatterOptions
        {
            JsonWriterOptions = new JsonWriterOptions() { Indented = false },
            IncludeScopes = true
        }
    );
    var logger = (ILogger)t.Logger;
    var sink = t.Sink;

    // Act
    using (logger.BeginScope("{Value}", value))
    {
        logger.LogInformation("{LogEntryValue}", value);
    }

    // Assert
    string message = sink.Writes[0].Message;
    AssertMessageValue(message, "Value");
    AssertMessageValue(message, "LogEntryValue");

    static void AssertMessageValue(string message, string propertyName)
    {
        var serializedValueMatch = Regex.Match(message, "\"" + propertyName + "\":(.*?),");
        Assert.Equal(2, serializedValueMatch.Groups.Count);
        string jsonValue = serializedValueMatch.Groups[1].Value;
        Assert.True(double.TryParse(jsonValue, NumberStyles.Any, CultureInfo.InvariantCulture, out var floatingPointValue), "The json doesn not contain a floating point value: " + jsonValue);
        Assert.Equal(1.2, floatingPointValue, 2);
    }
}
```
(It would require two separate regex patterns)
Thoughts?  Comments?

**Comment by danmoseley** [comment]
I agree I'd leave that one.

**Comment by kasperk81** [comment]
is it because Regex.Match() lighter than source generator here? or because it's test code?
two calls to Regex.Match() vs. two calls to Match() on generator returned object, i'd like if the latter won in both size and perf overall and code remain equally succinct.

**Comment by danmoseley** [comment]
@kasperk81 it's test code, clarity is more important than performance. Since we hardly at all use regex in our product (in this repo) changing to the source generator is essentially to test it.

**Comment by joperezr** [comment]
>Thoughts? Comments?

Agreed with @danmoseley. Source generator is great when you know the pattern already. For patterns that can change based on input not using the source generator is fine, even in this case when we kind of know what the values could be as this could be broken if the test is refactored.
